File: clipwright/schema/pipeline.py
"""Pipeline 数据模型 — 工作流执行上下文。

定义一次完整的视频生产管线的状态和执行轨迹。
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel, Field


class PipelineStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class PipelineStep(BaseModel):
    """管线中的单个执行步骤。"""
    agent_name: str
    status: PipelineStatus = PipelineStatus.PENDING
    started_at: Optional[datetime] = Field(default=None)
    completed_at: Optional[datetime] = Field(default=None)
    duration_ms: Optional[int] = Field(default=None)
    result: Optional[dict[str, Any]] = Field(default=None)
    error: Optional[str] = Field(default=None)
    retry_count: int = Field(default=0)


class PipelineRequest(BaseModel):
    """启动管线执行的请求。"""
    persona_id: str
    category_plugin_id: str
    topic: str
    extra_params: dict[str, Any] = Field(default_factory=dict)
    dry_run: bool = Field(default=False, description="仅生成预览，不渲染")
    use_v2: bool = Field(default=False, description="是否使用 v2 动态路由管线")


class PipelineState(BaseModel):
    """管线执行状态 — 完整的执行上下文。"""
    pipeline_id: str
    status: PipelineStatus = PipelineStatus.PENDING
    request: PipelineRequest
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)

    # 执行状态
    current_agent: Optional[str] = Field(default=None)
    steps: list[PipelineStep] = Field(default_factory=list)

    # Agent 间共享数据
    shared_data: dict[str, Any] = Field(default_factory=dict)

    # 最终输出
    output_path: Optional[str] = Field(default=None)
    error: Optional[str] = Field(default=None)

    def add_step(self, agent_name: str) -> PipelineStep:
        step = PipelineStep(agent_name=agent_name)
        self.steps.append(step)
        return step

    def get_step(self, agent_name: str) -> Optional[PipelineStep]:
        for s in self.steps:
            if s.agent_name == agent_name:
                return s
        return None

File: web/src/types/pipeline.ts
/**
 * Pipeline types — matching backend clipwright/schema/pipeline.py
 */

export type PipelineStatus =
  | 'pending'
  | 'running'
  | 'paused'
  | 'completed'
  | 'failed'
  | 'cancelled';

export type PipelinePhase =
  | 'idle'
  | 'structure'
  | 'material'
  | 'edit'
  | 'animation'
  | 'audio'
  | 'quality'
  | 'self_heal'
  | 'completed'
  | 'failed';

export interface PipelineStep {
  agent_name: string;
  status: PipelineStatus;
  started_at?: string | null;
  completed_at?: string | null;
  duration_ms?: number | null;
  result?: Record<string, unknown> | null;
  error?: string | null;
  retry_count: number;
}

export interface PipelineRequest {
  persona_id: string;
  category_plugin_id: string;
  topic: string;
  extra_params?: Record<string, unknown>;
  dry_run?: boolean;
  use_v2?: boolean;
}

export interface PipelineState {
  pipeline_id: string;
  status: PipelineStatus;
  request: PipelineRequest;
  created_at: string;
  updated_at: string;
  current_agent?: string | null;
  steps: PipelineStep[];
  shared_data: Record<string, unknown>;
  output_path?: string | null;
  error?: string | null;
}

/** SSE event types from pipeline trace stream */
export type PipelineSSEEventType =
  | 'agent_start'
  | 'agent_end'
  | 'agent_complete'
  | 'agent_error'
  | 'timeline_snapshot'
  | 'tool'
  | 'llm'
  | 'log'
  | 'info'
  | 'warning'
  | 'done'
  | 'self_heal'
  | 'pipeline_complete'
  | 'cancelled';

export interface AgentSuggestion {
  id: string;
  type: 'clip' | 'pace' | 'style' | 'transition';
  message: string;
  confidence: number;
  data?: Record<string, unknown>;
}

export interface AgentChatMessage {
  id: string;
  role: 'user' | 'assistant' | 'system';
  content: string;
  timestamp: string;
}

/** SSE log event types */
export type LogEventType =
  | 'agent_start'
  | 'agent_end'
  | 'llm'
  | 'tool'
  | 'skill'
  | 'plugin'
  | 'info'
  | 'warning'
  | 'error'
  | 'timeline_snapshot'
  | 'mg_start'
  | 'mg_end';

/** A single log entry from the pipeline execution log */
export interface LogEntry {
  id: string;
  timestamp: number;
  agent: string;
  type: LogEventType;
  summary: string;
  detail?: Record<string, unknown> | null;
  expanded?: boolean;
}

/** Per-agent timing stats */
export interface AgentStat {
  agent: string;
  startTime: number;
  endTime?: number;
  durationMs?: number;
}

/** Pipeline run summary */
export interface PipelineSummary {
  totalTokens: number;
  totalCost: string;
  materialCount: number;
  selfHealCount: number;
  agentDurations: Record<string, number>;
  timelineStats?: { tracks: number; clips: number; durationSec: number };
}
